File: kaf/net_canary.py
#!/usr/bin/env python3
"""KAF network canary (real module file, never run via -c/eval).

Exit 0 iff the network is UNREACHABLE, which is what a hermetic execution
cell requires. Exit 1 if any probe connects or resolves.
"""
import socket
import sys

reachable = []

for host, port in (("1.1.1.1", 443), ("8.8.8.8", 53)):
    try:
        s = socket.create_connection((host, port), timeout=2)
        s.close()
        reachable.append(f"tcp {host}:{port}")
    except OSError:
        pass

try:
    socket.getaddrinfo("one.one.one.one", 443)
    reachable.append("dns one.one.one.one")
except OSError:
    pass

if reachable:
    print("NETWORK=REACHABLE " + "; ".join(reachable))
    sys.exit(1)
print("NETWORK=NONE (all probes failed, hermetic)")
sys.exit(0)

File: kaf/net_canary.php
<?php
// KAF network canary (real program file, never run via php -r/eval).
// Exit 0 iff the network is UNREACHABLE (hermetic). Exit 1 otherwise.
$reachable = [];
foreach ([["1.1.1.1", 443], ["8.8.8.8", 53]] as $hp) {
    $errno = 0; $errstr = "";
    $fp = @fsockopen($hp[0], $hp[1], $errno, $errstr, 2.0);
    if ($fp !== false) { fclose($fp); $reachable[] = "tcp {$hp[0]}:{$hp[1]}"; }
    // A timeout (fsockopen sets errno 0) is INCONCLUSIVE, not proof of isolation:
    // fail closed. A real errno (ECONNREFUSED/ENETUNREACH/EHOSTUNREACH) IS isolation.
    elseif ($errno === 0) { $reachable[] = "tcp {$hp[0]}:{$hp[1]} (timeout: inconclusive, fail-closed)"; }
}
$ip = @gethostbyname("one.one.one.one");
if ($ip !== "one.one.one.one" && filter_var($ip, FILTER_VALIDATE_IP)) {
    $reachable[] = "dns one.one.one.one";
}
if (count($reachable) > 0) {
    echo "NETWORK=REACHABLE " . implode("; ", $reachable) . "\n";
    exit(1);
}
echo "NETWORK=NONE (all probes failed, hermetic)\n";
exit(0);
